services/src/machine_learning/ml_tasks.rs

```diff
@@ -0,0 +1,172 @@
+use crate::contexts::SessionContext;
+use crate::datasets::upload::UploadId;
+use crate::error;
+use crate::handlers::model_training::{
+    accumulate_raster_data, get_operators_from_workflows, get_query_processors,
+};
+use crate::tasks::{Task, TaskId, TaskManager, TaskStatusInfo};
+use crate::util::config::get_config_element;
+use crate::workflows::workflow::Workflow;
+use geoengine_datatypes::error::ErrorSource;
+use geoengine_datatypes::primitives::VectorQueryRectangle;
+use geoengine_datatypes::util::Identifier;
+use geoengine_operators::engine::ExecutionContext;
+use serde::{Deserialize, Serialize};
+use snafu::{ensure, ResultExt};
+use std::path::PathBuf;
+use std::sync::Arc;
+use tokio::fs;
+use utoipa::ToSchema;
+
+use super::{xgb_train_model, XgboostTrainingParams};
+
+use serde_json::Value;
+
+#[derive(Clone, Debug, Deserialize, Serialize, ToSchema)]
+// #[schema(example = json!({"input_workflows":[{"operator":{op1}},{"operator":{op2}}],"label_workflow":[{"operator":{op_lbl}],"params":{"aggregateVariant":"Simple","featureNames":["a","b","target"],"trainingConfig":{"eta":"0.75"}},"query":{"spatialBounds":{"lowerLeftCoordinate":{"x":-180.0,"y":-90.0},"upperRightCoordinate":{"x":180.0,"y":90.0}},"spatialResolution":{"x":1.0,"y":1.0},"timeInterval":{"end":1_385_899_200_000,"start":1_385_899_200_000}}}))]
+#[serde(rename_all = "camelCase")]
+pub struct MLTrainRequest {
+    pub params: XgboostTrainingParams,
+    pub input_workflows: Vec<Workflow>,
+    pub label_workflow: Vec<Workflow>,
+    pub query: VectorQueryRectangle,
+}
+
+/// parameter for the dataset from workflow handler (body)
+#[derive(Clone, Debug, Deserialize, Serialize, ToSchema)]
+#[schema(example = json!({"name": "foo", "description": null, "query": {"spatialBounds": {"upperLeftCoordinate": {"x": -10.0, "y": 80.0}, "lowerRightCoordinate": {"x": 50.0, "y": 20.0}}, "timeInterval": {"start": 1_388_534_400_000_i64, "end": 1_388_534_401_000_i64}, "spatialResolution": {"x": 0.1, "y": 0.1}}}))]
+#[serde(rename_all = "camelCase")]
+pub struct MachineLearningModelFromWorkflow {
+    pub name: String,
+    pub description: Option<String>,
+    pub query: VectorQueryRectangle,
+}
+
+/// response of the machine learning model from workflow handler
+#[derive(Clone, Debug, Deserialize, Serialize, ToSchema)]
+#[serde(rename_all = "camelCase")]
+pub struct MachineLearningModelFromWorkflowResult {
+    pub store_path: PathBuf,
+    pub model: Value,
+}
+
+impl TaskStatusInfo for MachineLearningModelFromWorkflowResult {}
+
+pub struct MachineLearningModelFromWorkflowTask<C: SessionContext> {
+    pub input_workflows: Vec<Workflow>,
+    pub label_workflow: Vec<Workflow>,
+    pub ctx: Arc<C>,
+    pub request: MLTrainRequest,
+    pub store_path: PathBuf,
+}
+
+impl<C: SessionContext> MachineLearningModelFromWorkflowTask<C> {
+    async fn process(&self) -> error::Result<MachineLearningModelFromWorkflowResult> {
+        let agg_variant = self.request.params.aggregate_variant.clone();
```

Comment:
do you really need to clone all these things? Some things like the aggregate variant could also be `Copy`, others you should be able to get away with borrowing it, right?

Reply:
do you need the training_config again? Otherwise you could just move it into the function?